broker/broker.go

```diff
@@ -151,22 +177,29 @@ func (b *Broker) Deprovision(
 		Details:    details,
 	}
 
-	operationData, isAsync, err := b.Provider.Deprovision(providerCtx, deprovisionData)
+	provisioner, ok := b.Provider.(provider.Provisioner)
+	if !ok {
+		return domain.DeprovisionServiceSpec{}, ErrNotImplemented
+	}
+	res, err := provisioner.Deprovision(providerCtx, deprovisionData)
 	if err != nil {
 		return domain.DeprovisionServiceSpec{}, err
 	}
+	if res == nil {
+		return domain.DeprovisionServiceSpec{}, ErrNilResponse
```

Comment:
if `nil` is not a valid response, can we change the `Provisioner` interface to not return a pointer? then we can ditch this and similar conditions.  It also looks more similar to the upstream ServiceBroker interface so people don't get confused when here it's pointers and elsewhere it's raw structs.

Reply:
the intention here is to make the return types nicer to work with. So here it is only be an error if the response is `nil` AND `err == nil` ....making it return a concrete type means all the error cases end up being:

```
return domain.ThingIDontCareAbout{}, errors.New("my error")
```

this isn't really wrong, but it can lead to some issues and in our case this actually caught a bug. we had some code like this...

```
if err != nil {
   return domain.Thing{}, nil
}
```

It should have been disguarding the Thing and returning the error... but because that is a totally valid return type, it went unoticed.

with the pointer return type...

```
if err != nil {
    return nil, nil
}
```

... would have stood out as wrong and would also have caused an error at runtime

Reply:
makes sense. Marking resolved.